drivers/include/mcp23017.h

```diff
@@ -0,0 +1,175 @@
+extern "C" {
+#endif
+
+#include "periph/i2c.h"
+#include "periph/gpio.h"
+
+/**
+ * @brief   Possible MCP23017 hardware addresses (wiring specific)
+ */
+enum {
+    MCP23017_ADDR_20 = 0x20, /**< I2C device address: A2=0, A1=0, A0=0 */
+    MCP23017_ADDR_21 = 0x21, /**< I2C device address: A2=0, A1=0, A0=1 */
+    MCP23017_ADDR_22 = 0x22, /**< I2C device address: A2=0, A1=1, A0=0 */
+    MCP23017_ADDR_23 = 0x23, /**< I2C device address: A2=0, A1=1, A0=1 */
+    MCP23017_ADDR_24 = 0x24, /**< I2C device address: A2=1, A1=0, A0=0 */
+    MCP23017_ADDR_25 = 0x25, /**< I2C device address: A2=1, A1=0, A0=1 */
+    MCP23017_ADDR_26 = 0x26, /**< I2C device address: A2=1, A1=1, A0=0 */
+    MCP23017_ADDR_27 = 0x27, /**< I2C device address: A2=1, A1=1, A0=1 */
+};
+
+/**
+ * @brief   Output Interrupt selection
+ */
+enum {
+    INTA,  /**< Output interrupt on INTA pin */
+    INTB   /**< Output interrupt on INTB pin */
+};
+
+/**
+ * @brief   Named return values
+ */
+enum {
+    MCP23017_OK          =  0,       /**< everything was fine */
+    MCP23017_NOI2C       = -1,       /**< I2C communication failed */
+    MCP23017_NODEV       = -2        /**< no MCP23017 device found on the bus */
+};
+
+/**
+ * @brief Device descriptor for the MCP23017 sensor
+ */
+typedef struct {
+    i2c_t i2c;                       /**< I2C device which is used */
+    uint8_t addr;                    /**< I2C address */
+    uint8_t port_a_dir;
+    uint8_t port_b_dir;
+    uint8_t port_a_pull_up;
+    uint8_t port_b_pull_up;
+    uint8_t port_a_value;
+    uint8_t port_b_value;
+    uint8_t port_a_inten;
+    uint8_t port_b_inten;
+} mcp23017_t;
+
+/**
+ * @brief   Struct containing the needed peripheral configuration
+ */
+typedef struct {
+    gpio_t int_a_pin;                /**< interrupt a pin */
+    gpio_t int_b_pin;                /**< interrupt b pin */
+} mcp23017_params_t;
+
+/**
+ * @brief   Initialize the MCP23017 driver.
+ *
+ * @param[out] dev          device descriptor of MCP23017 to initialize
+ *
+ * @return                  MCP23017_OK on success
+ * @return                  MCP23017_NOI2C if initialization of I2C bus failed
+ */
+int mcp23017_init(mcp23017_t *dev, const mcp23017_params_t* params);
+
+/**
+ * @brief   Read port of MCP23017
+ *
+ * @param[in]  dev          device descriptor of MCP23017
+ * @param[in]  port         device port
+ *
+ * @return                  value of port
+ * @return                  MCP23017_NODEV no device found
+ */
+int mcp23017_read_port(mcp23017_t *dev, uint8_t port);
+
+/**
+ * @brief   Set pin direction of MCP23017
+ *
+ * @param[in]  dev          device descriptor of MCP23017
+ * @param[in]  port         device port
+ * @param[in]  pin          device pin
+ * @param[in]  direction    pin direction
+ *
+ * @return                  MCP23017_OK on success
+ * @return                  MCP23017_NODEV no device found
+ */
+int mcp23017_set_dir(mcp23017_t *dev, uint8_t port, uint8_t pin, uint8_t dir);
```

Comment:
I would strongly suggest to rework the interface with the goal to make it more similar to the `periph/gpio.h` driver. As a first step, I think it would make sense to go with the following:
```c
int mcp23017_pin_init(mcp23017_t *dev, uint8_t pin, gpio_mode_t mode);
int mcp23017_pin_init_int(mcp_23017_t *dev, uint8_t pin, gpio_mode_t mode, gpio_flank_t flank, gpio_cb_t cb, void *arg);
int mcp23017_pin_read(mcp23017_t *dev, uint8_t pin);
int mcp23017_pin_set(mcp23017_t *dev, uint8_t pin);
...
```
The `pin` parameter as used above is used as abstraction for the 16 internal pins, so we simply map it `A0 -> 0, A1 -> 1, ..., B0 -> 8, B7 -> 15`. This needs of course to be documented in the module and API description for this driver...

As of now, we are not really able to map the gpio interface 1-on-1, so we have to live with some differences (e.g. set/clear/toggle have return values, as the i2c driver could fail).